adapters/jixie/jixie.go

```diff
@@ -0,0 +1,161 @@
+	bidder := &JixieAdapter{
+		endpoint: config.Endpoint,
+	}
+	return bidder, nil
+}
+
+// Adding header fields to request header
+func addHeaderIfNonEmpty(headers http.Header, headerName string, headerValue string) {
+	if len(headerValue) > 0 {
+		headers.Add(headerName, headerValue)
+	}
+}
+
+//func buildEndpoint(endpoint string, timeout int64) string {
+//	return endpoint + "?pstimeout=" + strconv.FormatInt(timeout, 10)
+//}
+
+func (a *JixieAdapter) MakeRequests(request *openrtb.BidRequest, reqInfo *adapters.ExtraRequestInfo) ([]*adapters.RequestData, []error) {
+	var errs []error
+
+	if len(request.Imp) == 0 {
+		return nil, []error{&errortypes.BadInput{
+			Message: fmt.Sprintf("No Imps in Bid Request"),
+		}}
+	}
+
+	for _, imp := range request.Imp {
+
+		var bidderExt adapters.ExtImpBidder
+		if err := json.Unmarshal(imp.Ext, &bidderExt); err != nil {
+			errs = append(errs, &errortypes.BadInput{
+				Message: fmt.Sprintf("Impression id=%s has an Error: %s", imp.ID, err.Error()),
+			})
+			continue
+		}
+		var jxExt openrtb_ext.ExtImpJixie
+		if err := json.Unmarshal(bidderExt.Bidder, &jxExt); err != nil {
+			errs = append(errs, &errortypes.BadInput{
+				Message: fmt.Sprintf("Impression id=%s, has invalid Ext", imp.ID),
+			})
+			continue
+		}
+	}
+
+	data, err := json.Marshal(request)
+	if err != nil {
+		errs = append(errs, err)
+	}
+
+	headers := http.Header{}
+	headers.Add("Content-Type", "application/json;charset=utf-8")
+	headers.Add("Accept", "application/json")
+	if request.Device != nil {
+		addHeaderIfNonEmpty(headers, "User-Agent", request.Device.UA)
+		addHeaderIfNonEmpty(headers, "X-Forwarded-For", request.Device.IP)
+		addHeaderIfNonEmpty(headers, "Accept-Language", request.Device.Language)
+		if request.Device.DNT != nil {
+			addHeaderIfNonEmpty(headers, "DNT", strconv.Itoa(int(*request.Device.DNT)))
+		}
+	}
+
+	if request.Site != nil {
+		addHeaderIfNonEmpty(headers, "Referer", request.Site.Page)
+	}
+
+	//theurl := buildEndpoint(a.endpoint, request.TMax)
+
+	return []*adapters.RequestData{{
+		Method:  "POST",
+		Uri:     a.endpoint,
+		Body:    data,
+		Headers: headers,
+	}}, errs
+}
+
+func ContainsAny(raw string, keys []string) bool {
+	lowerCased := strings.ToLower(raw)
+	for i := 0; i < len(keys); i++ {
+		if strings.Contains(lowerCased, keys[i]) {
+			return true
+		}
+	}
+	return false
+
+}
+
+func getBidType(bidAdm string) openrtb_ext.BidType {
+	if bidAdm != "" && ContainsAny(bidAdm, []string{"<?xml", "<vast"}) {
+		return openrtb_ext.BidTypeVideo
+	}
+	return openrtb_ext.BidTypeBanner
+}
+
+// MakeBids make the bids for the bid response.
+func (a *JixieAdapter) MakeBids(internalRequest *openrtb.BidRequest, externalRequest *adapters.RequestData, response *adapters.ResponseData) (*adapters.BidderResponse, []error) {
+
+	if response.StatusCode == http.StatusNoContent {
+		// no bid response
+		return nil, nil
+	}
```

Comment:
Other adapters return a `BadInput` error when a bad request is received. Would it be useful to add that scenario in our context?
```
124   func (a *JixieAdapter) MakeBids(internalRequest *openrtb.BidRequest, externalRequest *adapters.RequestData, response *adapters.ResponseData) (*adapters.BidderResponse, []error) {
125  
126       if response.StatusCode == http.StatusNoContent {
127           // no bid response
128           return nil, nil
129       }
    +
    +     if response.StatusCode == http.StatusBadRequest {
    +     	return nil, []error{&errortypes.BadInput{
    +     		Message: fmt.Sprintf("Unexpected status code: %d. Run with request.debug = 1 for more info", response.StatusCode),
    +     	}}
    +     }
130  
131       if response.StatusCode != http.StatusOK {
132           return nil, []error{&errortypes.BadServerResponse{
133               Message: fmt.Sprintf("Invalid Status Returned: %d. Run with request.debug = 1 for more info", response.StatusCode),
134           }}
135       }
adapters/jixie/jixie.go
```

Reply:
added. tks. in 0xaa2d1d5